python_modules/dagster/dagster/_core/workspace/load_target.py

```diff
@@ -41,24 +42,95 @@ def create_origins(self) -> Sequence[CodeLocationOrigin]:
+    module_name_present = "module_name" in dagster_block and isinstance(
+        dagster_block.get("module_name"), str
+    )
+    modules_present = "modules" in dagster_block and isinstance(
+        dagster_block.get("modules"), list
+    )
+
+    if module_name_present and modules_present:
+        # Here we have the check only for list; to be a bit more forgiving in comparison to 'is_valid_modules_list' in case it's an empty list next to 'module_name' existance
+        if len(dagster_block["modules"]) > 0:
+            raise ValueError(
+                "Only one of 'module_name' or 'modules' should be specified, not both."
+            )
+
+    if modules_present and len(dagster_block.get("modules")) == 0:
+        raise ValueError("'modules' list should not be empty if specified.")
+
+    return True
+
+
+def is_valid_modules_list(modules) -> bool:
+    # Could be skipped theorectically, but double check maybe useful, if this functions finds it's way elsewhere
+    if not isinstance(modules, list):
+        raise ValueError("Modules should be a list.")
+
+    for index, item in enumerate(modules):
+        if not isinstance(item, dict):
+            raise ValueError(f"Item at index {index} is not a dictionary.")
+        if "type" not in item:
+            raise ValueError(
+                f"Dictionary at index {index} does not contain the key 'type'."
+            )
+        if not isinstance(item["type"], str):
+            raise ValueError(
+                f"The 'type' value in dictionary at index {index} is not a string."
+            )
+        if "name" not in item:
+            raise ValueError(
+                f"Dictionary at index {index} does not contain the key 'name'."
+            )
+        if not isinstance(item["name"], str):
+            raise ValueError(
+                f"The 'name' value in dictionary at index {index} is not a string."
+            )
+
+    return True
+
+
+def get_origins_from_toml(
+    path: str,
+) -> Sequence[ManagedGrpcPythonEnvCodeLocationOrigin]:
     with open(path, "rb") as f:
         data = tomli.load(f)
         if not isinstance(data, dict):
             return []
 
         dagster_block = data.get("tool", {}).get("dagster", {})
+        origins = []
+
+        if "module_name" in dagster_block or "modules" in dagster_block:
+            assert validate_dagster_block_for_module_name_or_modules(dagster_block) is True
+
         if "module_name" in dagster_block:
-            return ModuleTarget(
-                module_name=dagster_block["module_name"],
+            origins += ModuleTarget(
+                module_name=dagster_block.get("module_name"),
                 attribute=None,
                 working_directory=os.getcwd(),
                 location_name=dagster_block.get("code_location_name"),
             ).create_origins()
-        return []
 
+        if "modules" in dagster_block and is_valid_modules_list(
```

Comment:
elif and direct returns would make it more clear that these causes are mutually exclusive